joy_teleop/joy_teleop/joy_teleop.py

```diff
@@ -251,71 +229,167 @@ def run_topic(self, c, joy_state):
+            return
+
+        goal = self.action_type.Goal()
+        for target, value in self.action_goal.items():
+            set_message_fields(goal, {target: value})
+
+        self.action_client.send_goal_async(goal)
+
+
+class JoyTeleop(Node):
+    """
+    Generic joystick teleoperation node.
+
+    Will not start without configuration, has to be stored in 'teleop' parameter.
+    See config/joy_teleop.yaml for an example.
+    """
+
+    def __init__(self):
+        super().__init__('joy_teleop', allow_undeclared_parameters=True,
+                         automatically_declare_parameters_from_overrides=True)
+
+        self.commands = []
+
+        names = []
+
+        for name, config in self.retrieve_config().items():
+            if name in names:
+                self.get_logger().error("command '{}' was duplicated".format(name))
+                raise JoyTeleopException("command '{}' was duplicated".format(name))
+
             try:
-                package, interface, message = type_name.split('/')
-                mod = importlib.import_module(package + ext)
-                self.message_types[type_name] = getattr(mod, message)
-            except ValueError:
-                raise JoyTeleopException('message type format error')
-            except ImportError:
-                raise JoyTeleopException('module {} could not be loaded'.format(package))
-            except AttributeError:
-                raise JoyTeleopException(
-                    'message {} could not be loaded from module {}'.format(package, message))
-        return self.message_types[type_name]
-
-    def update_actions(self, evt=None):
-        for name, cmd in self.command_list.items():
-            if cmd['type'] != 'action':
-                continue
-            if cmd['action_name'] in self.offline_actions:
-                self.register_action(name, cmd)
+                interface_group = config['type']
+
+                if interface_group == 'topic':
+                    self.commands.append(JoyTeleopTopicCommand(name, config, self))
+                elif interface_group == 'service':
+                    self.commands.append(JoyTeleopServiceCommand(name, config, self))
+                elif interface_group == 'action':
+                    self.commands.append(JoyTeleopActionCommand(name, config, self))
+                else:
+                    self.get_logger().error("unknown type '{interface_group}' "
+                                            "for command '{name}'".format_map(locals()))
+                    raise JoyTeleopException("unknown type '{interface_group}' "
+                                             "for command '{name}'".format_map(locals()))
+            except TypeError:
+                # This can happen on parameters we don't control, like
+                # 'use_sim_time'.  Just ignore it with a warning for the user.
+                self.get_logger().warn('parameter {} is not a dict'.format(name))
```

Comment:
I'm not exactly sure what you are thinking about here.  I guess we could make the assumption that if there are no dots in the parameter, it can't possibly be for us, but I think that's the only heuristic I would apply.  Other ones might have false positives.  Do you have other concrete suggestions?

Reply:
I presume a `TypeError` exception may get thrown because we blindly attempt to use every parameter value as a `dict`. If we only care about those that represent a `joy_teleop` config, why not ignoring the rest? I'd think we can check and drop them after looping in `retrieve_config()`. WDYT?